driver: sensor: Add APDS9960 sensor driver

punitvara opened the pull request:
This patch implements following functionality in polling mode.
- Ambient light
- RGB light
- proximity sensor

Apart from this it also adds Blue channel to channel type
to sensor header file.

Following datasheet has been used to develop driver
https://docs.broadcom.com/docs/AV02-4191EN

Jira: ZEP-1552

Signed-off-by: Punit Vara <punit.vara@intel.com>

MaureenHelm commented on drivers/sensor/apds9960/Kconfig:
```diff
@@ -0,0 +1,41 @@
+
+menuconfig APDS9960
+        bool
+        prompt "APDS9960 Sensor"
```
You can collapse lines 10-11 to:
``bool "APDS9960 Sensor"``

You can do this for all the other configs too.

punitvara replied:
Done.

MaureenHelm commented on drivers/sensor/apds9960/Kconfig:
```diff
@@ -0,0 +1,41 @@
+        default "APDS9960"
+        depends on APDS9960
+        help
+        Device name with which the APDS9960 is identified.
```
Indent two spaces

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/Kconfig:
```diff
@@ -0,0 +1,41 @@
+        help
+        Device name with which the APDS9960 is identified.
+
+config APDS9960_THREAD_PRIORITY
```
Remove this config, you're not using it anywhere

punitvara replied:
Done.

MaureenHelm commented on drivers/sensor/apds9960/Makefile:
```diff
@@ -0,0 +1,4 @@
+ccflags-y +=-I$(srctree)/drivers
+
+obj-$(CONFIG_APDS9960) += apds9960.o
+obj-$(CONFIG_APDS9960_TRIGGER) += apds9960_trigger.o
```
Remove this line

punitvara replied:
Done.

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,572 @@
+    /* Read current ENABLE register */
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS, 0x80 ,&enable_value) < 0){
+                SYS_LOG_DBG(" Failed");
+		return 0xFF;
```
Return an errno instead

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,572 @@
+    struct apds9960_data *data = dev->driver_data;
+
+    /* Read current ENABLE register */
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS, 0x80 ,&enable_value) < 0){
```
Replace 0x80 with APDS9960_ENABLE_REG

punitvara replied:
Done.

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,572 @@
+
+    /* Read current ENABLE register */
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS, 0x80 ,&enable_value) < 0){
+                SYS_LOG_DBG(" Failed");
```
Fix indentation

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,572 @@
+}
+
+bool apds9960_setmode(struct device *dev, uint8_t mode, uint8_t enable)
+{
```
This function seems unnecessarily complicated. Why not just use i2c_reg_update_byte instead?

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,572 @@
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
+			APDS9960_PDATA_REG, &pdata) < 0) {
+		return -EIO;
+	}
```
These registers are all contiguous so you should be able to replace all the i2c_reg_read_byte calls with one i2c_burst_read. Unless this sensor doesn't support address auto-increment?

punitvara replied:
I didn't find this auto- increment feature in datasheet

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,572 @@
+{
+	struct apds9960_data *data = dev->driver_data;
+
+	if (chan == SENSOR_CHAN_LIGHT) {
```
Let's use a switch here

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,572 @@
+        if (i2c_reg_write_byte(data->i2c, APDS9960_I2C_ADDRESS, APDS9960_CONTROL_REG,val) < 0){
+                SYS_LOG_DBG(" Failed");
+                return false;
+        }
```
Can simplify this function by using i2c_reg_update_byte instead

nashif commented:
@punitvara same story, before we even look at the code, you need to fix the style errors :)

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,572 @@
+
+	apds9960_sensor_setup(dev, als_gain);
+
+	dev->driver_api = &apds9960_driver_api;
```
Remove this line and use DEVICE_AND_API_INIT below instead

punitvara replied:
Done

MaureenHelm commented on include/sensor.h:
```diff
@@ -108,6 +108,8 @@ enum sensor_channel {
 	/** Illuminance in green spectrum, in lux. */
 	SENSOR_CHAN_GREEN,
+	/** Illuminance in blue spectrum, in lux. */
+	SENSOR_CHAN_BLUE,
```
Add this in a separate commit

punitvara replied:
Done

nashif reviewed (changes requested):
need a sample with this, please convert samples/sensor/apds9960/ to use this sensor directly

MaureenHelm reviewed (changes requested):
I think you can eliminate a few of the set* functions by calling i2c_reg_update_byte directly instead. This should also help reduce code size.

There are quite a few formatting issues throughout. Please address all the checkpatch errors.

All of the functions in apds9960.c need to be static.

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+		return false;
+	}
+
+	return false;
```
Shouldn't this be true?

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+	return true;
+}
+
+bool apds9960_enablepower(struct device *dev)
```
I think you could just return the value returned by the setmode function.

punitvara replied:
Done.

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+		return -EIO;
+	}
+
+	if (!(chip_id == APDS9960_ID_1) || (chip_id == APDS9960_ID_2)) {
```
As you wrote it, ! applies just for ID_1. I think you meant to do it for both of the IDs.

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+
+	if (i2c_reg_write_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_WTIME_REG, APDS9960_DEFAULT_WTIME) < 0) {
+		SYS_LOG_DBG("Defualt wait time not set ");
```
Typo - should be Default.

punitvara replied:
Done.

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+
+	if (i2c_reg_write_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_PPULSE_REG, APDS9960_DEFAULT_PROX_PPULSE) < 0) {
+		SYS_LOG_DBG("Defualt wait time not set ");
```
Wrong debug message (same for next ones).

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+	}
+
+	if (!apds9960_setmode(dev, APDS9960_ALL, APDS9960_MODE_OFF)) {
+		SYS_LOG_DBG("setmode not zero");
```
Update message to something more suggestive.

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+
+	if (i2c_reg_write_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_ATIME_REG, APDS9960_DEFAULT_ATIME) < 0) {
+		SYS_LOG_DBG("ADC Default integration time not set for ADC");
```
Remove fist word from message.

punitvara replied:
Done.

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+		return -EIO;
+	}
+
+	apds9960_proxy_setup(dev, gain);
```
Should check return values for these 2 setup functions.

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+	data->sample_b = 0;
+	data->pdata = 0;
+
+	apds9960_sensor_setup(dev, als_gain);
```
Check return value. Also, als_gain doesn't seem to initialized here.

punitvara replied:
If I initialize it to zero , checkpatch.pl threw error saying global variable is initialized

bogdan-davidoaia replied:
That is strange. I don't see why it would do that.

punitvara replied:
Anyway I initialized it with zero

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.h:
```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright (c) 2016 Intel Corporation
```
2017

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/apds9960/apds9960.h:
```diff
@@ -0,0 +1,279 @@
+#ifndef _SENSOR_APDS9960_H_
+#define _SENSOR_APDS9960_H_
+
+#include <device.h>
```
Do you really need all these includes here? I think it would be cleaner to keep just what is needed and move the rest to the .c file.

punitvara replied:
I have included all required header file here because only one header need to be added in .c file. Is it fine ?

nashif replied:
agree, printk is not being used in your code, so not sure why you are adding it here. The headers need to be included in the source file where they are using and and not via another header.

punitvara replied:
Done

bogdan-davidoaia commented on drivers/sensor/apds9960/Kconfig:
```diff
@@ -0,0 +1,41 @@
+menuconfig APDS9960
+        bool
+        prompt "APDS9960 Sensor"
+        depends on SENSOR && I2C && GPIO
```
Why require GPIO?

punitvara replied:
Removed

mshawcroft commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+
+#include "apds9960.h"
+
+u8_t apds9960_getmode(struct device *dev)
```
This and many of the following functions can be made static.

punitvara replied:
Done

mshawcroft commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+	return 0;
+}
+
+static struct sensor_driver_api apds9960_driver_api = {
```
These driver API structures should be const.

punitvara replied:
Done

mshawcroft commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,584 @@
+	return 0;
+}
+
+struct apds9960_data apds9960_data;
```
Looks like this structure can be static.

punitvara replied:
Done

punitvara commented on drivers/sensor/apds9960/Kconfig:
```diff
@@ -0,0 +1,41 @@
+        string
+        prompt "I2C Master"
+        depends on APDS9960
+        default "I2C_0"
```
Should I use just NAME or any related term instead of NAME ?

bogdan-davidoaia replied:
Just replace default "I2C_0" with default I2C_0_NAME.
 
I2C_0_NAME is a config for the I2C_0 name, so use that instead of the string directly.

punitvara replied:
Ok. thanks

nashif commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,591 @@
+
+	if (i2c_reg_write_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_PPULSE_REG, APDS9960_DEFAULT_PROX_PPULSE) < 0) {
+		SYS_LOG_DBG("Defualt proximity ppulse not set ");
```
Default

punitvara replied:
Done

nashif commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,591 @@
+
+	if (i2c_reg_write_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_POFFSET_UR_REG, APDS9960_DEFAULT_POFFSET_UR) < 0) {
+		SYS_LOG_DBG("Defualt poffset UR not set ");
```
Default, also below...

punitvara replied:
Done

mshawcroft commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,591 @@
+	return 0;
+}
+
+bool apds9960_setproxintlowthresh(struct device *dev, u8_t threshold)
```
This function, and I suspect, all of the following functions should be static.

mshawcroft commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,591 @@
+
+bool apds9960_setlightinthighthresh(struct device *dev, u16_t threshold)
+{
+struct apds9960_data *data = dev->driver_data;
```
The indentation here and in several other functions is wrong.

mshawcroft commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,591 @@
+	.channel_get = &apds9960_channel_get,
+};
+
+int apds9960_init(struct device *dev)
```
This definition should be static.

mshawcroft commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,591 @@
+		return -EIO;
+	}
+
+	if ((apds9960_ambient_setup(dev, gain)) < 0) {
```
apds_ambient_setup returns an errno hence this code should simply propagate the existing errno rather than replace it, I suspect the same comment applies to many of the preceding function calls but I've not checked them individually.

punitvara committed:
include: Add new sensor channel blue

This commit adds new sensor channel macro SENSOR_CHAN_BLUE which can
be used for RGB sensors to get illuminance in Blue spectrum.

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
drivers: Add apds9960 sensor in polling mode

This patch implements following functionality in polling mode.

Ambient light
RGB light
proximity sensor

Following datasheet has been used to develop driver
https://docs.broadcom.com/docs/AV02-4191EN

Jira: ZEP-1552

Signed-off-by: Punit Vara <punit.vara@intel.com>

punitvara committed:
samples: Add sample for APDS9960 sensor

Sample can be used to test ambient light, RGB and proximity
functionality of APDS9960 sensor in polling mode.

Remove old sample which is not using sensor API.

ZEP-1552

Signed-off-by: Punit Vara <punit.vara@intel.com>

dbkinder reviewed (changes requested):
I don't see any content in samples/sensor/2_apds9960/README.rst

punitvara commented:
Anything else  I have to do ? or everything looks fine to get this merge. ? @dbkinder

nashif commented on samples/sensor/2_apds9960/README.rst:
```diff
@@ -0,0 +1,36 @@
+.. _2_apds9960:
```
please pick a prober  name for sample, do not use numbers.. I would just remove the existing sample we have right now or make it use the new sensor.

nashif commented on samples/sensor/2_apds9960/testcase.ini:
```diff
@@ -0,0 +1,5 @@
+[test]
+build_only = true
+tags = samples
+arch_whitelist = arc
```
remove the arch_whitelist, it is not needed, this needs to work on all architectures.

nashif commented on samples/sensor/2_apds9960/README.rst:
```diff
@@ -0,0 +1,36 @@
+
+.. code-block:: console
+
+        proxy without trigger is -1
```
why we are getting -1?

punitvara replied:
Updated the output. proximity output depends on different situations

nashif commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,455 @@
+/* sensor_apds9960.c - driver for APDS9960 ALS/RGB/gesture/proximity sensor */
+
+/*
```
move this below the license header, also, use the @file notation instead.

punitvara replied:
Done

nashif commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,455 @@
+	return 0;
+}
+
+static bool apds9960_setproxintlowthresh(struct device *dev, u8_t threshold)
```
this function name is unreadable. Maybe add _ between words or shorten it.

punitvara replied:
Done

nashif commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,455 @@
+	return true;
+}
+
+static bool apds9960_setproxinthighthresh(struct device *dev, u8_t threshold)
```
ditto

mshawcroft replied:
This function and all of its friends squash the errno returns codes they get to true/false and the caller then converts false to EIO. Instead these functions should directly return the errno codes they catch rather than squashing to true/false, and the caller should likewise propagate the errno.

punitvara replied:
Done

nashif commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,455 @@
+	return true;
+}
+
+static bool apds9960_setlightintlowthresh(struct device *dev, u16_t threshold)
```
ditto

nashif commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,455 @@
+	return true;
+}
+
+static bool apds9960_setlightinthighthresh(struct device *dev, u16_t threshold)
```
ditto

nashif commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,455 @@
+	return true;
+	}
+
+static bool apds9960_setleddrive(struct device *dev, u8_t drive)
```
ditto

nashif reviewed (dismissed):
few inline comments

dbkinder commented on samples/sensor/2_apds9960/README.rst:
```diff
@@ -0,0 +1,36 @@
+.. _2_apds9960:
+
+APDS9960 RGB, ambient, Gesture Sensor
```
Suggested Title: `APDS-9960 Digital RGB and Gesture Sensor Sample`

dbkinder reviewed (changes requested):
How is this different from the existing https://www.zephyrproject.org/doc/samples/sensor/apds9960/README.html sample?  As Anas suggests, should/can these samples be combined and the doc for the current sample leveraged as needed?  The sample doc here is not helpful as written (compare with the existing apds9960 doc.)

mshawcroft commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,455 @@
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	/* Read lower byte following by MSB  Ref: Datasheet : RGBC register */
+	if (i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
```
All of these calls to i2c_ functions return errno's which should be returned rather than be replaced with EIO

punitvara replied:
Done

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,36 @@
 
-RGB and Gesture Sensor
-######################
+APDS9960 RGB, ambient, Gesture Sensor
```
APDS-9960 RGB and Gesture Sensor

punitvara replied:
According to datasheet https://cdn.sparkfun.com/assets/learn_tutorials/3/2/1/Avago-APDS-9960-datasheet.pdf

All three things has been mentioned and there for data registers to fetch data one for ambient light mean white light brightness and another three for RGB for IMHO this title provides more info than you suggested

dbkinder replied:
OK,  But then let's say "APDS9960 RGB, Ambient Light, and Gesture Sensor"

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,36 @@
-The APA102/C requires 5V data and clock signals, so logic level shifter
-(preferred) or pull-up resistors are needed.  Make sure the pins are 5V
-tolerant if using pull-up resistors.
+samples provide output to the console. It requires an apds9960 sensor and connect with
```
After connecting the APDS-9960 sensor to your Arduino 101 board (following the \`sparkfun APDS-9960 hookup guide\`_ for example), this sample code will write sensor data output to the serial console.  Use these commands to make and flash the sample to the board:

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,36 @@
 
-Building and Running
-********************
+        $ cd ../../..
```
you should explain what's happening here (there are two binaries being flashed...)

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,36 @@
-   $ make BOARD=arduino_101_sss flash # with JTAG
-   .. or
-   $ dfu-util -a sensor_core -D outdir/arduino_101_sss/zephyr.bin # with DFU
+        ambient light intensity without trigger is 387
```
explain what "without trigger" means?  Does the sample code just write the values out once and stop?  Does it query the sensor in loop?

punitvara replied:
Without trigger means without using interrupt in this case. I have already wrote in description that this sample code is just for polling mode. and of course polling means it will read sensor data every second.

dbkinder replied:
OK.  I've suggested edits above to clarify this, thanks.

punitvara replied:
Done

dbkinder reviewed (changes requested):
-1 doc needs more information about the sample and how to get it running (I took a guess at some hardware needed and pointed to a sparkfun example, but is that what you tested with?)

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,38 @@
-This sample utilizes APDS-9960 Sensor and reads RGB values from the sensor
-then displays the color through the APA102C LED.
+A sensor application that demonstrates how to use the APDS9960 to get ambient light,
+RGB and proximity data. This sample only checks sensor in polling mode(without trigger)
```
add a space before the parenthesis, and a final period. How about saying ...  "poling mode (without an interrupt trigger)."

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,38 @@
-Building and Running
-********************
+        $ cd ../../..
+        $ cd tests/booting/stub
```
Better would be to combine these two cd commands into one:

     $ cd $ZEPHYR_BASE/tests/booting/stub

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,38 @@
 
-   The APA102C are very bright even at low settings.  Protect your eyes
-   and do not look directly into those LEDs.
+        $ cd samples/sensors/apds9960
```
recommend instead:

     $ cd $ZEPHYR_BASE/samples/sensors/apds9960

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,38 @@
-The APA102/C requires 5V data and clock signals, so logic level shifter
-(preferred) or pull-up resistors are needed.  Make sure the pins are 5V
-tolerant if using pull-up resistors.
+samples provide output to the console. It requires an apds9960 sensor and 
```
This sample application uses an APDS9960 sensor connected to an Arduino 101 board
    (for example as shown in this `Sparkfun APDS9960 tutorial<https://www.sparkfun.com/products/12787>`__.
    There are two processor cores (x86 and ARC) on the Arduino 101.  You'll need to flash
    both stub code on the x86 core and this sample's code on the ARC core, as explained below.

dbkinder reviewed (changes requested):
a few clarity and grammar edits...

punitvara commented:
@dbkinder  Is this ready to merge ?

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,40 @@
+`Sparkfun APDS9960 tutorial<https://www.sparkfun.com/products/12787>`.
+There are two processor cores (x86 and ARC) on the Arduino 101.  You'll need to flash
+both this sample's code on the ARC core (using the `arduino101_sss` board target,
+and stub code on the x86 core (using the `arduino_101` board target), as show below:
```
as shown below

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,40 @@
-tolerant if using pull-up resistors.
+This sample application uses an APDS9960 sensor connected to an Arduino 101 board
+(for example as shown in this
+`Sparkfun APDS9960 tutorial<https://www.sparkfun.com/products/12787>`.
```
add the closing paren   ")"

dbkinder reviewed (changes requested):
a few formatting tweaks and we're good

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,515 @@
+		val->val1 = data->sample_c;
+		val->val2 = 0;
+		break;
+	case (SENSOR_CHAN_RED):
```
Why are SENSOR_CHAN_RED and friends wrapped in parens?

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,515 @@
+
+	switch (chan) {
+	case SENSOR_CHAN_LIGHT:
+		val->val1 = data->sample_c;
```
Does the raw data need to be converted to standard units?

punitvara replied:
No.

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,515 @@
+	int temp = 0;
+
+	/* Read value from CONTROL register */
+	temp = i2c_reg_read_byte(data->i2c, APDS9960_I2C_ADDRESS,
```
You can use i2c_reg_update_byte instead

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,515 @@
+
+	/* ADC value */
+	temp = i2c_reg_update_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_ATIME_REG, APDS9960_ATIME_WRTIE, 0xB6);
```
Where does 0xB6 come from?

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,515 @@
+
+	/* proxy Gain */
+	temp = i2c_reg_update_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_CONTROL_REG, (BIT(2) | BIT(3)), (gain & 0x0C));
```
Use a macro to define the mask

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,515 @@
+
+	/* Enable proxy */
+	temp = i2c_reg_update_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_ENABLE_REG, (BIT(2) | BIT(3)), 0x05);
```
Use macros to define the mask and value. Is 0x05 the right value since you haven't set BIT(0) in the mask?

punitvara replied:
Done

MaureenHelm commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,40 @@
-then displays the color through the APA102C LED.
+This sample application demonstrates how to use the APDS9960 sensor to get
+ambient light, RGB, and proximity (or gesture) data.  This sample checks the
+sensor in poling mode (without an interrupt trigger).
```
s/poling/polling/

punitvara replied:
Didn't get you. Can you please elaborate ?

punitvara replied:
Done

MaureenHelm commented on samples/sensor/apds9960/prj.conf:
```diff
@@ -2,4 +2,9 @@ CONFIG_STDOUT_CONSOLE=y
-CONFIG_I2C_0_DEFAULT_CFG=0x14
+CONFIG_PRINTK=y
+CONFIG_APDS9960=y
+CONFIG_APDS9960_TRIGGER_NONE=y
```
remove this line

punitvara replied:
Done

MaureenHelm reviewed (changes requested):
There are several magic numbers in the driver that need to be fixed.

Is there a build test for this sample?

punitvara commented on drivers/sensor/apds9960/Makefile:
```diff
@@ -0,0 +1,3 @@
+ccflags-y +=-I$(srctree)/drivers
```
Done

punitvara commented:
" Is there a build test for this sample? " I didn't get you. @dbkinder @MaureenHelm let me know if I need to do any change to get this merge.

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,489 @@
+
+	/* proxy Gain */
+	temp = i2c_reg_update_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_CONTROL_REG, APDS9960_BITS_32,
```
APDS9960_BITS_32 needs a more descriptive name, such as APDS9960_CONTROL_PGAIN. Same for the other bitfields

punitvara replied:
Done.

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,489 @@
+	}
+
+	temp = i2c_reg_update_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_ENABLE_REG, BIT(5), 0x00);
```
This transaction and the one above it both update the same register. Can you combine them into one?

punitvara replied:
Later when we introduce interrupt we need to write this bit on some conditions( #ifdef) so let not combine this as of now.

punitvara replied:
Does everything everything look fine ?

MaureenHelm replied:
Unless the hardware requires otherwise, I think you should still do this in one transaction.

MaureenHelm replied:
In other words, do the #ifdef-ery to figure out the value to write before you call `i2c_reg_update_byte()`. Something like:
```c
u8_t mask = APDS9960_ENABLE_PROXY | APDS9960_ENABLE_PIEN;
u8_t val = APDS9960_PROXY_ON;
#ifdef CONFIG_APDS9960_INTERRUPT
val |= APDS9960_ENABLE_PIEN;
#endif

/* Configure the PON, PEN, and PIEN fields in the ENABLE register */
temp = i2c_reg_update_byte(data->i2c, APDS9960_I2C_ADDRESS,
 		APDS9960_ENABLE_REG, mask, value);
```

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,489 @@
+
+	/* ALE Gain */
+	temp = i2c_reg_update_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_CONTROL_REG, APDS9960_BITS_10,
```
Here's another case where a more descriptive macro name is needed. Replace APDS9960_BITS_10 with APDS9960_CONTROL_AGAIN

punitvara replied:
Done.

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,489 @@
+
+	/* Enable ALE */
+	temp = i2c_reg_update_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_ENABLE_REG, APDS9960_BITS_10, APDS9960_RGB_ON);
```
Can this transaction to the same register be combined with the one before it?

punitvara replied:
Later when we introduce interrupt we need to write this bit on some conditions( #ifdef) so let not combine this as of now. What do you think is it ready to merge ?

MaureenHelm reviewed (changes requested):
> " Is there a build test for this sample? " I didn't get you.

I want to make sure this driver gets built when you run sanitycheck. Looks like it's already done, in samples/sensor/apds9960/sample.yaml

punitvara commented:
@nashif  I have cleared all the review comments. Can you please check whether this is ready to merge or  not ?

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,42 @@
+`Sparkfun APDS9960 tutorial`_).
+There are two processor cores (x86 and ARC) on the Arduino 101.  You'll need to flash
+both this sample's code on the ARC core (using the `arduino101_sss` board target),
+and stub code on the x86 core (using the `arduino_101` board target), as show below:
```
change "as show below" to "as shown below"

dbkinder replied:
need two back ticks:

     using the ``arduino_101`` board target

punitvara replied:
Done

punitvara replied:
Done

punitvara replied:
@dbkinder let me know if we can merge this because this is for 1.9

dbkinder commented on samples/sensor/apds9960/README.rst:
```diff
@@ -1,52 +1,42 @@
+(for example as shown in this
+`Sparkfun APDS9960 tutorial`_).
+There are two processor cores (x86 and ARC) on the Arduino 101.  You'll need to flash
+both this sample's code on the ARC core (using the `arduino101_sss` board target),
```
need two back ticks here:

    using the ``arduino101_sss``

dbkinder replied:
need two back ticks:

     using the ``arduino101_sss``

punitvara replied:
Done

dbkinder reviewed (changes requested):
still a few edits

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,489 @@
+	temp = i2c_reg_write_byte(data->i2c, APDS9960_I2C_ADDRESS,
+		APDS9960_PILT_REG, threshold);
+	if (temp < 0) {
+		SYS_LOG_DBG(" Failed");
```
Replace SYS_LOG_DBG with SYS_LOG_ERR here and throughout

punitvara replied:
Done

MaureenHelm commented on drivers/sensor/apds9960/apds9960.c:
```diff
@@ -0,0 +1,489 @@
+
+	temp = apds9960_proxy_setup(dev, gain);
+	if (temp < 0) {
+		SYS_LOG_DBG("Failed to setup proxymity functionality");
```
s/proxymity/proximity/

punitvara replied:
For above feedback. I wrote interrupt bit separately because later I want to implement 
#if INTERRUPT_IS_ON
write interrupt bit 1
#else 
write interrupt bit 0

That's why I haven't combined with above transaction. All other sensor driver do the same.

MaureenHelm reviewed (changes requested):
> Here's another case where a more descriptive macro name is needed. Replace APDS9960_BITS_10 with APDS9960_CONTROL_AGAIN

Thanks for fixing this

dbkinder reviewed (dismissed):
+1 for docs

dbkinder reviewed (approved):
still +1 for docs
